server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -128,28 +129,16 @@ public static long sizeOfUnaccountedShards(
 
         // Where reserved space is unavailable (e.g. stats are out-of-sync) compute a conservative estimate for initialising shards
         for (ShardRouting routing : node.initializing()) {
-            if (routing.relocatingNodeId() == null && metadata.getIndexSafe(routing.index()).isSearchableSnapshot() == false) {
-                // in practice the only initializing-but-not-relocating non-searchable-snapshot shards with a nonzero expected shard size
-                // will be ones created
-                // by a resize (shrink/split/clone) operation which we expect to happen using hard links, so they shouldn't be taking
-                // any additional space and can be ignored here
-                continue;
-            }
-            if (reservedSpace.containsShardId(routing.shardId())) {
-                continue;
-            }
-            final String actualPath = clusterInfo.getDataPath(routing);
-            // if we don't yet know the actual path of the incoming shard then conservatively assume it's going to the path with the least
-            // free space
-            if (actualPath == null || actualPath.equals(dataPath)) {
-                totalSize += getExpectedShardSize(
-                    routing,
-                    Math.max(routing.getExpectedShardSize(), 0L),
-                    clusterInfo,
-                    null,
-                    metadata,
-                    routingTable
-                );
+            if (shouldReserveSpaceForInitializingShard(routing, metadata) && reservedSpace.containsShardId(routing.shardId()) == false) {
```

Comment:
I _think_ I have seen a situation when reserved space was actually smaller then a size of the shard when adding a new replica. We should consider changing `ReservedSpace` record to include per shard sizes and update below computation to do `max(reservedShardSize, expectedShardSize, getExpectedShardSize(..))`

Reply:
This require additional investigation. Possibly there is either a race condition or we simply see somewhat outdated ClusterInfo

Reply:
Maybe you can have a ticket or something https://github.com/elastic/elasticsearch/issues/91386 to remember to tackle this.

Reply:
Okay, I checked and it behaves as expected. Reserved space should only account for the size to be downloaded (smaller than the actual shard size). The rest is already accounted in the current disk usage.